File: LC/Medium/102.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        q = []
        q.append(root)
        ans = []
        
        while q:
            curr_level = []
            
            for i in range(len(q)):
                curr = q.pop(0)
                if curr:
                    curr_level.append(curr.val)
                
                    if curr.left:
                        q.append(curr.left)
                    if curr.right:
                        q.append(curr.right)
            
            if curr_level:
                ans.append(curr_level)
        
        return ans

File: LC/C++/Medium/102.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
 #include <queue>
class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        if (!root) {
            return vector<vector<int>>{};
        }
        vector<vector<int>> ans;
        queue<TreeNode*> q;
        q.push(root);

        while (!q.empty()) {
            vector<int> currLevel;
            int n = q.size();

            for (int i = 0; i < n; ++i) {
                TreeNode* currNode = q.front();
                q.pop();
                currLevel.emplace_back(currNode->val);

                if (currNode->left) {
                    q.push(currNode->left);
                }

                if (currNode->right) {
                    q.push(currNode->right);
                }
            }
            ans.emplace_back(currLevel);
        }

        return ans;
    }
};